torch/csrc/autograd/python_variable.cpp

```diff
@@ -150,8 +150,9 @@ static PyObject* THPVariable_make_subclass(PyObject* _ignored, PyObject* args, P
   if (!PyType_Check(cls)) {
     throw TypeError("cls must be a type (got %s)", Py_TYPE(cls)->tp_name);
   }
-  auto data = as_variable_ref(r.tensor(1)).tensor_data();
-  auto var = make_variable(data, r.toBool(2));
+  auto data = as_variable_ref(r.tensor(1)).variable_data();
+  data.unsafeGetTensorImpl()->set_allow_tensor_metadata_change(true);
```

Comment:
We set `data`'s `allow_tensor_metadata_change` to true here, because we want to allow the following use case:
```python
rnn = torch.nn.RNN(100, 100, 2)
rnn.cuda()  # this internally calls `rnn.flatten_parameters()`, which calls `torch._cudnn_rnn_flatten_weight(...)`, which is expected to change storage of `rnn`'s weights in-place 
```
for backward compatibility.

Reply:
Two points:

1) Is there any reason this can't use re-use the version counter, i.e. `detach` semantics rather than `data`?

2) If we are going with `detach` semantics, passing a boolean to allow the metadata change seems reasonable -- have we not done a release yet where the "allow_metadata_change" is in?

Reply:
From in-person discussion:
1. We should use `detach` instead of `variable_data` here, because the original method `tensor_data` shares version counter with the source tensor, which is the same semantics as `detach`.
2. We should explicitly set the `allow_tensor_metadata_change` flag to true here, instead of adding the `allow_tensor_metadata_change` flag to `detach()`, because there is not much of a need to have `detach(allow_tensor_metadata_change=True)` outside of the `Variable()` and `nn.Parameter()` constructors in Python.